plugins/ansilexer.py

```diff
@@ -28,83 +29,288 @@
+        'ffff00',
+        '0000ff',
+        'ff00ff',
+        '00ffff',
+        'ffffff',
+    ]
+    _palette_cube_steps = ['00', '5f', '87', 'af', 'd7', 'ff']
+
+    def __init__(self, **options):
+        RegexLexer.__init__(self, **options)
+
+        self._bright = False
+        self._foreground = 'Default'
+        self._background = 'Default'
+
+    def _callback(self, match):
+        commands = match.group('commands')
+        text = match.group('text')
+
+        # split the commands strings into their constituent parameter codes
+        parameters = self._sgr_split.split(commands)[1:-1]
+        parameters = [int(p) if p else 0 for p in parameters]
+
+        # loop over the parameters, consuming them to create commands, some
+        # of which will have arguments
+        while parameters:
+            command = parameters.pop(0)
+            if command == 0:
+                self._bright = False
+                self._foreground = 'Default'
+                self._background = 'Default'
+            elif command == 1:
+                self._bright = True
+            elif command == 22:
+                self._bright = False
+            elif command >= 30 and command <= 37:
+                self._foreground = self._named_colors[command - 30]
+            elif command == 38:
+                mode = parameters.pop(0)
+                if mode == 2:
+                    rgb = self._read_rgb(parameters)
+                else:
+                    offset = parameters.pop(0)
+                    rgb = self._make_color_from_palette(offset)
+                self._foreground = 'ForegroundColor' + rgb
+            elif command == 39:
+                self._foreground = 'Default'
+            elif command >= 40 and command <= 47:
+                self._background = self._named_colors[command - 40]
+            elif command == 48:
+                mode = parameters.pop(0)
+                if mode == 2:
+                    rgb = self._read_rgb(parameters)
+                else:
+                    offset = parameters.pop()
+                    rgb = self._make_color_from_palette(offset)
+                self._background = rgb
+            elif command == 49:
+                self._background = 'Default'
+            elif command >= 90 and command <= 97:
+                self._foreground = ('Bright' +
+                                    self._named_colors[command - 90])
+            elif command >= 100 and command <= 107:
+                self._background = ('Bright' +
+                                    self._named_colors[command - 100])
+
+        if (self._bright
+            and not self._foreground.startswith('Bright')
+            and not self._foreground.startswith('ForegroundColor')):
+            token = 'Generic.AnsiBright' + self._foreground
         else:
+            token = 'Generic.Ansi' + self._foreground
+
+        if (self._background != 'Default'):
+            token += 'BackgroundColor' + self._background
```

Comment:
The BackgroundColor section of the token name has a different convention than the foreground, as this prefix always appears, to distinguish from the foreground color. I assumed that the representation of the tokens wasn't a huge deal, so long as the formatter spits out the right thing…